DEP: switch default of special.comb to legacy=False

j-bowhay opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
#15722
#### What does this implement/fix?
<!--Please explain your changes.-->
As per the issue, change the default of `legacy` to False and raise a warning if `legacy=True`.
#### Additional information
<!--Any additional information you think is important.-->
@h-vetinari would be good if you could check if this was what you had in mind.

h-vetinari commented on scipy/special/_basic.py:
```diff
@@ -2602,12 +2602,10 @@ def comb(N, k, exact=False, repetition=False, legacy=True):
         arguments is unaffected by the value of `exact`.
 
         .. deprecated:: 1.9.0
```
There's some more text to change in the description of the `exact` argument

h-vetinari commented on scipy/special/_basic.py:
```diff
@@ -2639,22 +2637,20 @@ def comb(N, k, exact=False, repetition=False, legacy=True):
+        )
     if repetition:
         return comb(N + k - 1, k, exact, legacy=legacy)
     if exact:
         if int(N) != N or int(k) != k:
```
This should now be redone; the conditions were only there to make the warning less noisy
```suggestion
    if exact:
        if int(N) == N and int(k) == k:
            # _comb_int casts inputs to integers, which is safe & intended here
            return _comb_int(N, k)
        elif legacy:
            # here at least one number is not an integer; legacy behavior just casts to int
            return _comb_int(N, k)
        # otherwise, we disregard `exact=True`; it makes no sense for non-integral arguments
        return comb(N, k)
    else:
```

h-vetinari commented on scipy/special/tests/test_basic.py:
```diff
@@ -1326,17 +1326,13 @@ def test_comb(self):
             ):
                 result = special.comb(N, k, exact=True, legacy=legacy,
                                       repetition=repetition)
-        else:
-            result = special.comb(N, k, exact=True, legacy=legacy,
-                                  repetition=repetition)
-        if legacy:
```
Aside from the change in the warning, this test shouldn't have to change AFAICT.

h-vetinari commented on scipy/special/_basic.py:
```diff
@@ -2639,22 +2637,20 @@ def comb(N, k, exact=False, repetition=False, legacy=True):
     220
 
     """
+    if legacy:
+        warnings.warn(
+            "Using 'legacy=True' is deprecated and will raise an error by "
+            "Scipy 1.13.0. If you want to keep the legacy behaviour, cast "
```
We're only showing this warning if someone sets `legacy=True`, but if someone has added `legacy=False`, they will not see the warning and - AFAIU - will be broken by the removal in 1.13. I think the warning here is fine, but we should not error on `legacy=False` in 1.13

j-bowhay replied:
Thanks for the review @h-vetinari. Do you seen any reason we can't set the default to `None` for this release then have seperate warnings for `legacy=True` and `legacy=False` so that we can remove `legacy` by 1.13?

h-vetinari replied:
It's possible. My point in #15722 was that we shouldn't punish those most diligent to respond to the original deprecation warning (by adding `legacy=False`) by showing a warning again right away. Bug I guess it's a fairly niche problem and people will just move on.

h-vetinari reviewed (commented):
Thanks for opening this PR and for pinging me!

j-bowhay commented:
Thanks for the review @h-vetinari. I believe I have addressed your comments so would be grateful if you could take another look. I have changed the deprecation slightly so that we could remove `legacy` faster.

h-vetinari reviewed (approved):
Thanks! :)

h-vetinari commented on scipy/special/_basic.py:
```diff
@@ -2639,25 +2636,28 @@ def comb(N, k, exact=False, repetition=False, legacy=True):
+            # here at least one number is not an integer; legacy behavior uses
+            # lossy casts to int
+            return _comb_int(N, k)
+        # otherwise, we disregard `exact=True`; it makes no sense for
+        # non-integral arguments
+        return comb(N, k)
```
We could start warning about using `exact=True` with non-integers, but I'm fine to discuss this in a separate issue/PR.

tupui reviewed (approved):
LGTM, I will let you merge @h-vetinari if it's fine for you too.

h-vetinari commented:
Thanks @j-bowhay 🥳
